### PR TITLE
feat(processor): add support for workspace-level isolation

### PR DESCRIPTION
**Pull request by atzoum**
# Description

Introducing support for workspace isolation in the processor module.

## Design Doc
[Link](https://www.notion.so/rudderstacks/Design-Proposal-Processor-Isolation-2db79f6c5b0f4a9fb14465b62aaaf54a)

Untitled

### `utils` package
- `misc.ExponentialNumber` is used by processor to  sleep for exponentially increasing durations while no jobs are being picked up.
- `queue.PriorityQueue` implements a priority queue pretty close to the standard one [here](https://pkg.go.dev/container/heap#example-package-PriorityQueue).
- `sync.Limiter` provides a concurrency limiter which supports dynamic priorities (i.e. priority increases periodically as time goes by for avoiding starvation of lower priority routines) and captures useful statistics. It is inspired by [vivek-ng/concurrency-limiter](https://github.com/vivek-ng/concurrency-limiter).

### `processor` package
- `isolation.Strategy` provides both the interface and the 2 different isolation strategy implementations. This shall be the only extension point for introducing additional isolation strategies in the future.
- `processor.worker`  implements the processor worker as described in the design doc.
- `processor.workerPool` implements the worker pool as described in the design doc.
- `processor.workerHandle` is the interface that abstracts `processor.Handle` implementation from the worker by including only the required methods.
- `processor.Handle#Start` starts the pinger loop as described in the design doc.

## Workspace Priorities

By default, all workspace have the same (low) priority with respect to their chance of getting picked up by the various concurrency limiters (get jobs, preprocess, transform, store) when there is a waiting queue. It is possible though to treat a workspace with higher priority, by setting the `Processor.Limiter.<WorkspaceID>.Priority` configuration property to a value between 2-4:
- 1: low priority
- 2: medium priority
- 3: medium/high priority
- 4: high priority

To make sure that low-priority workspaces will not be starved, every second, their priority will be increasing by 1 until it reaches to 4.

## Testing & Benchmarks

To ease testing of multi-workspace deployments (a.k.a multi-tenant) without the need for a properly seeded etcd server, support has been added for forcing the use of a static `ModeProvider` through the `forceStaticModeProvider` config property.

Workspace isolation overall behaviour and performance testing is performed in `processor_isolation_test.go`.  In comparison to the "no isolation" mode, workspace isolation:
- Performs **equally well** with up to **200 active** workspaces. 
- With **500 active** workspaces it is **18% slower**.
- With **1000 active** workspaces it is **30% slower**. 
- With **10000 active** workspaces it is **150% slower**.

To provide some perspective, in our current production workloads, the average number of active workspaces is not greater than 50 & the same number for destinations is between100-150.

A visualisation dashboard which demonstrates both the new statistics that are being captured, as well as the server's performance during the benchmarks can be found [here](https://snapshots.raintank.io/dashboard/snapshot/9643HY7OqrsYae1HuAjAz167953NOCN3).

## Notion Ticket

[Link](https://www.notion.so/rudderstacks/8aac9087df644365acdf64e28e290153?v=6e06b0a5ade24f0aa5ffe05dc2972e84&p=1bd87345dc654c26ab2fb9e92d98f7df&pm=s)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: testhelper/destination/transformer.go [hunk 1]
```diff
@@ -21,13 +20,13 @@ func SetupTransformer(pool *dockertest.Pool, d cleaner) (*TransformerResource, e
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to pull image: %w", err)
-	}
+	// err := pool.Client.PullImage(docker.PullImageOptions{
+	// 	Repository: "rudderlabs/rudder-transformer",
+	// 	Tag:        "latest",
+	// }, docker.AuthConfiguration{})
+	// if err != nil {
+	// 	return nil, fmt.Errorf("failed to pull image: %w", err)
+	// }
```
Comment (atzoum) [inline testhelper/destination/transformer.go:29]: need to revert this change

File: jobsdb/jobsdb.go [hunk 2]
```diff
@@ -1307,6 +1313,12 @@ func (jd *HandleT) createDSInTx(tx *Tx, newDS dataSetT) error {
 		return err
 	}
 
+	sqlStatement = fmt.Sprintf(`CREATE INDEX "idx_%[1]s_ws" ON %[1]q (workspace_id)`, newDS.JobTable)
+	_, err = tx.ExecContext(context.TODO(), sqlStatement)
+	if err != nil {
+		return err
+	}
```
Comment (atzoum) [inline jobsdb/jobsdb.go:1320]: migrator needs to add the same index when creating tables on its own

File: app/apphandlers/embeddedAppHandler.go [hunk 3]
```diff
@@ -62,6 +63,7 @@ func (a *embeddedApp) loadConfiguration() {
 	config.RegisterIntConfigVariable(0, &a.config.gatewayDSLimit, true, 1, "Gateway.jobsDB.dsLimit", "JobsDB.dsLimit")
 	config.RegisterIntConfigVariable(0, &a.config.routerDSLimit, true, 1, "Router.jobsDB.dsLimit", "JobsDB.dsLimit")
 	config.RegisterIntConfigVariable(0, &a.config.batchRouterDSLimit, true, 1, "BatchRouter.jobsDB.dsLimit", "JobsDB.dsLimit")
+	config.RegisterBoolConfigVariable(false, &a.config.forceStaticMode, false, "forceStaticModeProvider")
```
Comment (atzoum) [inline app/apphandlers/embeddedAppHandler.go:66]: **_Note:_** 
For testing purposes, we need to be able to support both a static ModeProvider and a Multi-tenant DeploymentType both at the same time, so that we don't need to do the _"etcd dance"_ when testing multi-workspace scenarios.

`processor_isolation_test.go` is such an example

File: jobsdb/jobsdb.go [hunk 4]
```diff
@@ -1911,6 +1923,53 @@ func (jd *HandleT) GetPileUpCounts(ctx context.Context) (map[string]map[string]i
...
+			(SELECT s.* FROM t, LATERAL(
+			  SELECT workspace_id FROM %[1]q f
+			  WHERE f.workspace_id > t.workspace_id
+			  ORDER BY workspace_id LIMIT 1) s)
+		  )
+		  SELECT * FROM t) a`, ds.JobTable))
+		} else {
+			queries = append(queries, fmt.Sprintf(`SELECT DISTINCT workspace_id from %[1]q j LEFT JOIN "v_last_%[2]s" js ON js.job_id = j.job_id
+			WHERE js.job_state is NULL or js.job_state not in ('aborted', 'succeeded', 'migrated')`, ds.JobTable, ds.JobStatusTable))
+		}
```
Comment (atzoum) [inline jobsdb/jobsdb.go:1954]: _**Note:**_ 
To assess what is more valuable, a fast query or a precise one, taking into consideration [this](https://www.notion.so/rudderstacks/Design-Proposal-Processor-Isolation-2db79f6c5b0f4a9fb14465b62aaaf54a#8bb2901bd4374bdfaaf30d2ca823b0e3), along with the fact that jobsdb's cache is not enabled for the `gw` rightmost dataset

File: jobsdb/jobsdb.go (hunk 2 above)
Comment (atzoum) [reply jobsdb/jobsdb.go:1320]: https://github.com/rudderlabs/rudder-migrator/pull/35

File: jobsdb/jobsdb.go (hunk 4 above)
Comment (fracasula) [reply jobsdb/jobsdb.go:1954]: What do you think about being redundant with the workspace_id and have it in the status table too and hit just that index without a join? I think I would've find it useful to have it there a few times.

Comment (fracasula) [reply jobsdb/jobsdb.go:1954]: Also, what is the performance difference between the two? Is it negligible with the number of workspaces that we deal with?

File: utils/misc/exponential.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	expo.value = 0
+}
+
+// Next returns the next number, which is the previous one multiplied by 2, always abiding by the min and max provided.
+func (expo *ExponentialNumber[T]) Next(min, max T) T {
+	expo.value *= 2
+	if expo.value > max {
+		expo.value = max
+	}
+	if expo.value < min {
```
Comment (fracasula) [inline utils/misc/exponential.go:27]: With <=1 the function will always return `min` e.g.:

```go
var exp misc.ExponentialNumber[float64]
t.Log(exp.Next(0.9, 1)) // prints 0.9 since 0.9^2 = 0.81 which is less than min
```

I'm pointing it out only because it is a generic component and it can be used by anyone. A power of a <1 is still possible, it's just that it decreases.
Given the intended use for this, perhaps we should complain if that is attempted. Either way it shouldn't be a big deal.

File: utils/queue/priorityQueue.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	Value     T
+	Priority  int
+	timeStamp int64
+	index     int
+}
+
+// PriorityQueue ....
+type PriorityQueue[T any] []*Item[T]
+
+// Len: Size of the priority queue . Used to satisfy the heap interface...
```
Comment (fracasula) [inline utils/queue/priorityQueue.go:19]: I got some linting stuff out of this file. Beside JetBrains complaining for `:` after `Len` saying:
> Comment should have the following format 'Len ...' (with an optional leading article) 

On a more serious note, some methods of `PriorityQueue` have method receivers, some do not. Is that intended?

File: utils/queue/priorityQueue.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+// Item stores the attributes which will be pushed to the priority queue..
+type Item[T any] struct {
+	Value     T
+	Priority  int
+	timeStamp int64
+	index     int
+}
+
+// PriorityQueue ....
```
Comment (fracasula) [inline utils/queue/priorityQueue.go:16]: So meaningful :laughing:

File: utils/queue/priorityQueue.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+// Pop removes elements from the priority queue
+func (pq *PriorityQueue[T]) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	if n == 0 {
+		return nil
+	}
+	item := old[n-1]
+	old[n-1] = nil
```
Comment (fracasula) [inline utils/queue/priorityQueue.go:55]: Is this line necessary given that we'll reslice `pq`? :thinking:

File: utils/queue/priorityQueue.go (hunk 7 above)
Comment (fracasula) [reply utils/queue/priorityQueue.go:16]: Given that as you put it it's very close to the heap std implementation, I think it would be worth explaining in a comment how it is different from a `std` one and when to use this implementation.

File: utils/sync/limiter.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	l.mu.Unlock()
+
+	// no dynamic priority
+	if l.dynamicPeriod == 0 || priority == LimiterPriorityValueHigh {
+		<-w.Value
+		return
+	}
+
+	// dynamic priority (increment priority every dynamicPeriod)
+	ticker := time.NewTicker(l.dynamicPeriod)
```
Comment (fracasula) [inline utils/sync/limiter.go:174]:
```suggestion
	ticker := time.NewTicker(l.dynamicPeriod)
	defer ticket.Stop()
```

File: utils/sync/limiter_test.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+		statsTriggerCh := make(chan time.Time)
+		triggerFn := func() <-chan time.Time {
+			return statsTriggerCh
+		}
+
+		limiter := miscsync.NewLimiter(ctx, &wg, "test", 1, ms, miscsync.WithLimiterStatsTriggerFunc(triggerFn))
+		var counter int
+		statsTriggerCh <- time.Now()
+		statsTriggerCh <- time.Now()
```
Comment (fracasula) [inline utils/sync/limiter_test.go:29]: How come we need to trigger twice?

File: processor/worker.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+// newWorker creates a new worker
+func newWorker(ctx context.Context, partition string, h workerHandle) *worker {
+	w := &worker{
+		handle:    h,
+		logger:    h.logger().Child(partition),
+		partition: partition,
+	}
+	w.lifecycle.ctx, w.lifecycle.cancel = context.WithCancel(ctx)
+	w.channel.ping = make(chan struct{}, 1)
+	w.channel.preproccess = make(chan subJob, w.handle.config().pipelineBufferedItems)
```
Comment (fracasula) [inline processor/worker.go:25]: Typo in `proccess`. It is repeated a few times.

File: jobsdb/jobsdb.go (hunk 4 above)
Comment (atzoum) [reply jobsdb/jobsdb.go:1954]: Overall performance difference is negligible when table count is low. It becomes more prominent when table count grows. With the number of workspaces we are having it should be safer to use the index-only scan, since we can tolerate false positives better than large table counts.

With respect to the status table, it is not enough on its own, since the status table doesn't hold entries for unprocessed events, which we need to be aware of.

File: utils/misc/exponential.go (hunk 5 above)
Comment (atzoum) [reply utils/misc/exponential.go:27]: The algorithm is not raising the previous number to any power, it is merely multiplying the previous number by 2. Adding another test case using the scenario you mentioned for clarity
```
	t.Run("exponential float", func(t *testing.T) {
		var exp misc.ExponentialNumber[float64]
		require.EqualValues(t, 0.9, exp.Next(0.9, 5))
		require.EqualValues(t, 1.8, exp.Next(0.9, 5))
		require.EqualValues(t, 3.6, exp.Next(0.9, 5))
		require.EqualValues(t, 5, exp.Next(0.9, 5))
		require.EqualValues(t, 5, exp.Next(0.9, 5))
		exp.Reset()
		require.EqualValues(t, 0.9, exp.Next(0.9, 5))
	})
```

File: utils/queue/priorityQueue.go (hunk 6 above)
Comment (atzoum) [reply utils/queue/priorityQueue.go:19]: Yes, using pointer and non-pointer receivers per case is intended and not any different than [Go's own proposed implementation](https://pkg.go.dev/container/heap#example-package-PriorityQueue) of a priority queue.

File: utils/queue/priorityQueue.go (hunk 7 above)
Comment (atzoum) [reply utils/queue/priorityQueue.go:16]: There is no std implementation of a priority queue, at least none that I am aware of. There is only an [example](https://pkg.go.dev/container/heap#example-package-PriorityQueue) in the docs. The only standard is the `heap.Interface` that types must implement to be used for heap operations.

File: utils/sync/limiter_test.go (hunk 10 above)
Comment (atzoum) [reply utils/sync/limiter_test.go:29]: To make sure that the first operation that was triggered has been completed

Comment (atzoum) [reply utils/sync/limiter_test.go:29]: To make sure that the first operation that was triggered has been completed

File: utils/queue/priorityQueue.go (hunk 8 above)
Comment (atzoum) [reply utils/queue/priorityQueue.go:55]: According to the [example](https://pkg.go.dev/container/heap#example-package-PriorityQueue)
```go
old[n-1] = nil  // avoid memory leak
item.index = -1 // for safety
```

**Review by fracasula — APPROVED** [review]
LGTM :+1:

File: processor/worker.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+		defer w.logger.Debugf("ping loop stopped for worker: %s", w.partition)
+		for {
+			select {
+			case <-w.lifecycle.ctx.Done():
+				return
+			case <-w.channel.ping:
+			}
+
+			w.setIdleSince(time.Time{})
+			if w.pickJobs() {
```
[the remark this answers is not in the document]
Comment (atzoum) [reply processor/worker.go:77]: Yes but I found it risky to introduce (i.e. respect) context in this step, since all our processes, so far, have been designed using a panic-driven approach, meaning that they cannot tolerate interruptions well.

**Review by Sidddddarth — APPROVED** [review]

File: jobsdb/jobsdb.go [hunk 13]
```diff
@@ -1307,6 +1313,12 @@ func (jd *HandleT) createDSInTx(tx *Tx, newDS dataSetT) error {
 		return err
 	}
 
+	sqlStatement = fmt.Sprintf(`CREATE INDEX "idx_%[1]s_ws" ON %[1]q (workspace_id)`, newDS.JobTable)
```
[the remark this answers is not in the document]
Comment (atzoum) [reply jobsdb/jobsdb.go:1316]: Keeping both indexes until we stop supporting an isolation mode, e.g. if we stop supporting workspace isolation this index could be dropped.

[the remark this answers is not in the document]
Comment (atzoum) [reply jobsdb/jobsdb.go:1316]: Just realized that for existing tables we are missing this index, so I am adding a migration script 👍
